### PR TITLE
update buffers to use Tokio 0.3 MPSC channels

### PR DESCRIPTION
**Pull request by hawkw**
This branch updates `linkerd2-buffer`, and `linkerd2-proxy-discover`'s
`buffer` module to use Tokio 0.3's MPSC channel rather than Tokio 0.2's.
The rest of the proxy still uses Tokio 0.2, including the 0.2 runtime.

Most of the Tokio synchronization primitives lost their `poll`-based
interfaces in 0.3 as part of the move to intrusive lists of wakers for
synchronization primitives (see tokio-rs/tokio#2325,
tokio-rs/tokio#2509, and tokio-rs/tokio#2861). This change takes
advantage of the inherently pinned nature of `async fn` and `async`
blocks to avoid needing a separate heap allocation to store the waiter
state for a task waiting on a synchronization primitive. However, it
means that a synchronization primitive can _only_ be waited on when the
future that waits on it is pinned --- otherwise, there is a potential
dangling pointer. The `poll`-based APIs allowed waiting on
synchronization primitives from unpinned contexts, so they were removed.

To wait on the synchronization primitives from contexts that may not be
pinned, such as `poll_ready`, it's necessary to add a `Pin<Box<...>>`
around the future that's waiting on the synchronization primitive. This
ensures that the future will not move while it's part of the wait list.
It's important to note that this isn't an _additional_ allocation per
waiter versus Tokio 0.2; instead, it's the same allocation that would
have _always_ happened internally to the synchronization primitive in
the 0.2 API. Now, it's moved outside of the `tokio::sync` type so that
it can be avoided when used with `async`/`await` syntax, and added by
the user when polling the sync primitives.

Because we need to poll channel senders in `tower::Service`
implementations' `poll_ready` functions, it was necessary to introduce
our own bounded MPSC channel type that exposes a polling-based API. When
the buffer's channel is full, we want to exert backpressure in
`poll_ready`, so that callers such as load balancers could choose to
call another service rather than waiting for buffer capacity. This
branch adds a new `linkerd2-channel` crate that implements a pollable
bounded channel, wrapping `tokio::sync`'s unbounded MPSC and using a
`tokio::sync::Semaphore` to implement bounding. It's worth noting that
this is, essentially, how `tokio::sync::mpsc`'s bounded channel is
implemented --- it also uses the semaphore. However, our implementation
exposes a `poll_ready` method by boxing the future that waits to acquire
a semaphore permit, which the Tokio channel does not expose.

Finally, I've added some tests for the `linkerd2-channel` crate, based
on Tokio's tests for the MPSC channel, modified where the APIs differ.
This should help ensure we get similar behavior to what we expect from
Tokio's MPSCs.

This was factored out of PR #732.

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: linkerd/buffer/src/dispatch.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub(crate) async fn run<S, Req, I>(
             req = requests.recv().fuse() => {
                 match req {
                     None => return,
-                    Some(InFlight { request, tx }) => {
+                    Some(InFlight { request, tx, .. }) => {
```
Comment (olix0r) [inline linkerd/buffer/src/dispatch.rs:32]: nit: Are there other fields in `InFlight` now?
```suggestion
                    Some(InFlight { request, tx }) => {
```

File: linkerd/buffer/src/dispatch.rs [hunk 2]
```diff
@@ -44,7 +44,7 @@ pub(crate) async fn run<S, Req, I>(
                                 while let Some(InFlight { tx, .. }) = requests.recv().await {
                                     let _ = tx.send(Err(error.clone().into()));
                                 }
-                                return;
+                                break;
```
Comment (olix0r) [inline linkerd/buffer/src/dispatch.rs:47]: tioli: but `return` is clearer to me that it's leaving the function scope. `break` makes me have to inspect the parent scopes to figure out whether there are inner loops,e tc.

**Review by olix0r — APPROVED** [review]

File: linkerd/buffer/src/dispatch.rs (hunk 1 above)
Comment (hawkw) [reply linkerd/buffer/src/dispatch.rs:32]: no, there was a permit here before the channel was factored out to also be used in `discovery::buffer`, but i forgot to change this back --- good eye!

File: linkerd/buffer/src/dispatch.rs (hunk 2 above)
Comment (hawkw) [reply linkerd/buffer/src/dispatch.rs:47]:
```suggestion
                                return;
```

File: linkerd/buffer/src/dispatch.rs [hunk 3]
```diff
@@ -54,7 +54,7 @@ pub(crate) async fn run<S, Req, I>(
             e = idle().fuse() => {
                 let error = ServiceError(Arc::new(e.into()));
                 trace!(%error, "Idling out inner service");
-                return;
+                break;
```
Comment (olix0r) [inline linkerd/buffer/src/dispatch.rs:57]:
```suggestion
                return;
```

Comment (hawkw) [reply linkerd/buffer/src/dispatch.rs:57]: way ahead of you, buddy https://github.com/linkerd/linkerd2-proxy/pull/759/commits/e2046e304e1beb7fb36911f3612765f9a597cc5e :D

Comment (olix0r) [reply linkerd/buffer/src/dispatch.rs:57]: @hawkw i think that got the other break, but there were two.. so i think this is still relevant?

**Pull request merged**